src/inference/Inference.jl

```diff
@@ -77,24 +84,368 @@ function mh_accept(H::T, H_new::T, log_proposal_ratio::T) where {T<:Real}
+
+    # Chain construction.
+    return Chains(
+        convert(Array{Real}, parray),
+        string.(nms),
+        deepcopy(TURING_INTERNAL_VARS);
+        evidence=le,
+        info=info
+    )
+end
+
+function save(c::Chains, spl::AbstractSampler, model, vi, samples)
+    nt = NamedTuple{(:spl, :model, :vi, :samples)}((spl, model, deepcopy(vi), samples))
+    return setinfo(c, merge(nt, c.info))
+end
+
+function resume(c::Chains, n_iter::Int; kwargs...)
+    @assert !isempty(c.info) "[Turing] cannot resume from a chain without state info"
+    return sample(
+        c.info[:range],
+        c.info[:model],
+        c.info[:spl],
+        n_iter;    # this is actually not used
+        resume_from=c,
+        reuse_spl_n=n_iter,
+        kwargs...
+    )
+end
+
+function set_resume!(
+    s::Sampler;
+    resume_from::Union{Chains, Nothing}=nothing,
+    kwargs...
+)
+    # If we're resuming, grab the sampler info.
+    if resume_from != nothing
+        s = resume_from.info[:spl]
+    end
+end
+
+function split_var_str(var_str)
+    ind = findfirst(c -> c == '[', var_str)
+    inds = Vector{String}[]
+    if ind == nothing
+        return var_str, inds
+    end
+    sym = var_str[1:ind-1]
+    ind = length(sym)
+    while ind < length(var_str)
+        ind += 1
+        @assert var_str[ind] == '['
+        push!(inds, String[])
+        while var_str[ind] != ']'
+            ind += 1
+            if var_str[ind] == '['
+                ind2 = findnext(c -> c == ']', var_str, ind)
+                push!(inds[end], strip(var_str[ind:ind2]))
+                ind = ind2+1
+            else
+                ind2 = findnext(c -> c == ',' || c == ']', var_str, ind)
+                push!(inds[end], strip(var_str[ind:ind2-1]))
+                ind = ind2
+            end
+        end
+    end
+    return sym, inds
+end
+
+#########################
+# Default sampler state #
+#########################
+
+"""
+A blank `AbstractSamplerState` that contains only `VarInfo` information.
+"""
+mutable struct SamplerState{VIType<:VarInfo} <: AbstractSamplerState
+    vi :: VIType
+end
+
+#######################################
+# Concrete algorithm implementations. #
+#######################################
+
 include("hmc.jl")
 include("mh.jl")
 include("is.jl")
 include("AdvancedSMC.jl")
 include("gibbs.jl")
+include("../contrib/inference/sghmc.jl")
+include("../contrib/inference/AdvancedSMCExtensions.jl")
+
+################
+# Typing tools #
+################
 
-getspace(::IS) = ()
-getspace(::Type{<:IS}) = ()
-for alg in (:SMC, :PG, :PMMH, :IPMCMC, :MH)
+for alg in (:SMC, :PG, :PMMH, :IPMCMC, :MH, :IS)
     @eval getspace(::$alg{space}) where {space} = space
     @eval getspace(::Type{<:$alg{space}}) where {space} = space
 end
 for alg in (:HMC, :HMCDA, :NUTS, :SGLD, :SGHMC)
     @eval getspace(::$alg{<:Any, space}) where {space} = space
     @eval getspace(::Type{<:$alg{<:Any, space}}) where {space} = space
 end
+getspace(::Gibbs) = Tuple{}()
```

Comment:
Why is this not the union of the spaces of the algs? And if one of the algs has space `()`, then this should be `()` too.

Reply:
Because `Gibbs` always needs to access every parameter (and will likely always need to do so), and constructing the union of spaces seems like more complexity than is necessary, without adding much value.

Reply:
Ok in that case `()` is the same as `Tuple{}()`.